Bip 5

juanjux opened the pull request:
Signed-off-by: Juanjo Alvarez <juanjo@sourced.tech>

dennwc commented:
Reviewed 12 of 215 files at r1.
Review status: 12 of 215 files reviewed at latest revision, all discussions resolved, some commit checks broke.

---

*[driver/normalizer/annotation.go, line 15 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdpIz6AmCaNk_x1GNb:-LCdpIz6AmCaNk_x1GNc:b-k23ino) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L15)):*
> ```Go
> 	{
> 		ResponseMetadata{
> 			TopLevelIsRootNode: true,
> ```

You may omit the whole `ResponseMetadata` helper if this parameter is `true`.

---

*[driver/normalizer/annotation.go, line 62 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdp_AADoTWT97TS0o-:-LCdp_AADoTWT97TS0o0:b-8yg57p) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L62)):*
> ```Go
> 	}
> 
> 	if gostr, ok := s.Native().(string); ok {
> ```

same: `string(s)` instead of `s.Native().(string)`
and please check other places as well

---

*[driver/normalizer/annotation.go, line 102 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdptaM5PxZR0iXWl23:-LCdptaM5PxZR0iXWl24:boj5i98) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L102)):*
> ```Go
> 	}
> 
> 	if _, ok := operatorRoles[s]; !ok {
> ```

This is exactly what `Lookup` does, but I haven't added it to the guide yet :P

---

*[driver/normalizer/annotation.go, line 357 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdqpdd8be9Tz9491Xt:-LCdqpdd8be9Tz9491Xu:b-ctfans) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L357)):*
> ```Go
> 	// Parent of the last element of the qualified identifier (annotates
> 	// it and the child which is a normal identifier)
> 	MapAST("send", Obj{
> ```

Looks like a good place to use `AnnotateType` (it supports nesting now!).

---

*[driver/normalizer/annotation.go, line 374 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdqQJGA0SP3K4RTYaa:-LCdqQJGA0SP3K4RTYab:b3fatw7) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L374)):*
> ```Go
> 			"selector": Var("childselector"),
> 			uast.KeyRoles: Roles(role.Identifier),
> 			"__notcall": Int(1),
> ```

`Bool(true)`?

---

*[fixtures/var_global.rb.uast, line 8 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdsHLe4oTuOQOg2zlH:-LCdsHLe4oTuOQOg2zlI:b-2t96hp) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/fixtures/var_global.rb.uast#L8)):*
> ```
> .  Children: {
> .  .  0: module {
> .  .  .  Roles: Identifier,Module,Statement
> ```

`Identifier`?

---

*[fixtures/while.rb.uast, line 5 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdrT8-ES3O1chOzDTD:-LCdrT8-ES3O1chOzDTE:bl4tcg3) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/fixtures/while.rb.uast#L5)):*
> ```
> Errors: 
> UAST: 
>  {
> ```

Looks like we now have an empty node as a root

---

*[fixtures/while.rb.uast, line 30 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdrn0Y22SO_ktpwMek:-LCdrn0Y22SO_ktpwMel:b-4axbsc) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/fixtures/while.rb.uast#L30)):*
> ```
> .  .  .  .  .  Children: {
> .  .  .  .  .  .  0: true {
> .  .  .  .  .  .  .  Roles: Boolean,Condition,Expression,Literal
> ```

missing `While` role, compared to an old AST

---

*[fixtures/while.rb.uast, line 47 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCds5AZBgVU5ueRdUfh:-LCds5AZBgVU5ueRdUfi:boahfw3) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/fixtures/while.rb.uast#L47)):*
> ```
> .  .  .  .  .  .  }
> .  .  .  .  .  .  1: lvasgn {
> .  .  .  .  .  .  .  Roles: Assignment,Binary,Body,Expression,Identifier,Left
> ```

same for `While` role

---

*[native/lib/ruby_driver/node_converter.rb, line 78 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdrEfd8UywOiW-oQ2z:-LCdrEfd8UywOiW-oQ3-:b72jp77) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/native/lib/ruby_driver/node_converter.rb#L78)):*
> ```Ruby
>       # the inner nodes of the above
>       when "Complex", "Rational", "Symbol"
>         return {@@typekey => node_type(node), "token" => node.to_s}
> ```

`@token`?

---

*[native/lib/ruby_driver/node_converter.rb, line 232 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdoovt-V7JNm9JOdY8:-LCdoovt-V7JNm9JOdY9:b-t0cnl6) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/native/lib/ruby_driver/node_converter.rb#L232)):*
> ```Ruby
>       subelem = node.loc.send(key)
>       if subelem != nil
>         hash["pos_line_start"] = subelem.begin.line
> ```

Maybe we can use new names for position fields here as well?

---


*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/ruby-driver/21)*
<!-- Sent from Reviewable.io -->

juanjux commented:
Review status: 11 of 215 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[driver/normalizer/annotation.go, line 374 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdqQJGA0SP3K4RTYaa:-LChDiUR7RpvdG4_KYI4:b-896fix) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L374)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

`Bool(true)`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/ruby-driver/21)*
<!-- Sent from Reviewable.io -->

juanjux commented:
Review status: 11 of 215 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[driver/normalizer/annotation.go, line 374 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdqQJGA0SP3K4RTYaa:-LChDlmgBJP5l5-oXwRT:b-nc1h9m) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L374)):*
<details><summary><i>Previously, juanjux (Juanjo Alvarez Martinez) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Will probably review it anyway, my spider sense tells me there must be a better way (the "send" annotation is pretty complicated, but only the one for functions without parameters is failing, I'll need to check with you because probably finding the solution is a good test for the new DSL).

---


*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/ruby-driver/21)*
<!-- Sent from Reviewable.io -->

juanjux commented:
Review status: 11 of 215 files reviewed at latest revision, 11 unresolved discussions, some commit checks broke.

---

*[fixtures/var_global.rb.uast, line 8 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdsHLe4oTuOQOg2zlH:-LChEGUmCF3VPoflk-6G:b-q2l6uh) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/fixtures/var_global.rb.uast#L8)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

`Identifier`?
</blockquote></details>

Done. I'm not sure why reviewable still marks this as pending...

---


*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/ruby-driver/21)*
<!-- Sent from Reviewable.io -->

juanjux commented:
Review status: 9 of 226 files reviewed at latest revision, 7 unresolved discussions.

---

*[driver/normalizer/annotation.go, line 15 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdpIz6AmCaNk_x1GNb:-LCnm7Xm4jyhquxoZyVb:b-896fix) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L15)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

You may omit the whole `ResponseMetadata` helper if this parameter is `true`.
</blockquote></details>

Done.

---

*[fixtures/while.rb.uast, line 5 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdrT8-ES3O1chOzDTD:-LCnmCGZ8at3EG6-k-pd:b-896fix) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/fixtures/while.rb.uast#L5)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Looks like we now have an empty node as a root
</blockquote></details>

Done.

---

*[native/lib/ruby_driver/node_converter.rb, line 78 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdrEfd8UywOiW-oQ2z:-LCnmDHD9ldYak6Tbz-c:b-896fix) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/native/lib/ruby_driver/node_converter.rb#L78)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

`@token`?
</blockquote></details>

Done.

---

*[native/lib/ruby_driver/node_converter.rb, line 232 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdoovt-V7JNm9JOdY8:-LCnmEC1AKiFG8onuOxT:b-hyyxfo) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/native/lib/ruby_driver/node_converter.rb#L232)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Maybe we can use new names for position fields here as well?
</blockquote></details>

Those are actually the names defined in legacy.go but, as talked on Slack, if I remove the ObjectToNoder mapping them they don't work.

---


*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/ruby-driver/21)*
<!-- Sent from Reviewable.io -->

juanjux commented:
Review status: 5 of 226 files reviewed at latest revision, 7 unresolved discussions.

---

*[driver/normalizer/annotation.go, line 102 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdptaM5PxZR0iXWl23:-LDgZlWa2Qb3qvhP3HTQ:b-7w2vfj) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L102)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

This is exactly what `Lookup` does, but I haven't added it to the guide yet :P
</blockquote></details>

Fixed! (and a lot of cruft removed).

---

*[driver/normalizer/annotation.go, line 357 at r1](https://reviewable.io/reviews/bblfsh/ruby-driver/21#-LCdqpdd8be9Tz9491Xt:-LDgZqHo4wlSY1HPjcxD:b-4s1w3y) ([raw file](https://github.com/bblfsh/ruby-driver/blob/d120500f315d22d950938b8f7af37776da4932d5/driver/normalizer/annotation.go#L357)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Looks like a good place to use `AnnotateType` (it supports nesting now!).
</blockquote></details>

Changed all that part with the last commit.

---


*Comments from [Reviewable](https://reviewable.io/reviews/bblfsh/ruby-driver/21)*
<!-- Sent from Reviewable.io -->

dennwc commented on native/exe/native:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/ruby -wU
+#!/usr/bin/ruby -wUI/opt/driver/src/build/dependencies
```
This won't work locally, right?

juanjux replied:
Mmmm, I don't remember why I did that, but both `make test` and rebuilding the docker image work without it, I'll remove it and see if CI works here.

dennwc commented on native/lib/ruby_driver/node_converter.rb:
```diff
@@ -272,20 +306,20 @@ def add_comments()
-          "end_line" => comment.loc.last_line,
-          "start_col" => comment.loc.column,
-          "end_col" => comment.loc.last_column
+          "pos_line_start" => comment.loc.first_line,
```
Same here

juanjux commented on native/lib/ruby_driver/node_converter.rb:
```diff
@@ -224,15 +226,16 @@ def sexp_to_hash(node, table, cdr_index=nil, cdr_key=nil)
-        hash["end_line"] = subelem.end.line
-        hash["start_col"] = subelem.begin.column
-        hash["end_col"] = subelem.end.column
+        hash["pos_line_start"] = subelem.begin.line
```
Aren't these the new names already?

dennwc replied:
These are variable names, it uses `uast.KeyStart` in a separate node with a type `ast:Position`.

juanjux replied:
So something like: `hash["ast:Position"] = {"start": bla, "end": bla, "offset": bla}?`

juanjux replied:
Just saw the right format in the Java driver. Will change it.

juanjux replied:
Done, works, :+1:

dennwc commented:
LGTM! Thanks!
